src/v/transform/logging/record_batcher.cc

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "transform/logging/record_batcher.h"
+
+#include "absl/algorithm/container.h"
+#include "model/record.h"
+#include "storage/record_batch_builder.h"
+#include "transform/logging/logger.h"
+#include "utils/human.h"
+#include "utils/vint.h"
+
+namespace transform::logging {
+
+namespace detail {
+class batcher_impl {
+public:
+    batcher_impl() = delete;
+    explicit batcher_impl(size_t batch_max_bytes)
+      : _batch_max_bytes(batch_max_bytes) {}
+    ~batcher_impl() = default;
+    batcher_impl(const batcher_impl&) = delete;
+    batcher_impl& operator=(const batcher_impl&) = delete;
+    batcher_impl(batcher_impl&&) = delete;
+    batcher_impl& operator=(batcher_impl&&) = delete;
+
+    model::record_batch make_batch_of_one(iobuf k, iobuf v) {
+        return std::move(bb_init().add_raw_kv(std::move(k), std::move(v)))
+          .build();
+    }
+
+    void append(iobuf k, iobuf v) {
+        auto tmp = make_batch_of_one(std::move(k), std::move(v));
+        // A conservative estimate for the ultimate size of this record
+        size_t record_size = tmp.data().size_bytes();
+        record_size += vint::vint_size(static_cast<int64_t>(record_size));
```

Comment:
Why append this size? This seems like it's assuming that it's making a `model::record` object instead, which it's `size_bytes` excludes the record size itself. https://github.com/redpanda-data/redpanda/blob/de0eb9f266db076770ad42c6f16ff964e92088f3/src/v/model/record.h#L165-L166

Maybe it's because we need a better constructor that calculates all this information.

Reply:
>Why append this size?

Basically to pessimize the overall size estimate. If I go by `size_bytes`, the estimated batch size seems to pretty consistently undershoot the size of the final batch. What's not clear to me is whether the unaccounted for overhead is predictable/structured OR maybe an artifact of appending odd-sized data to the `iobuf` behind `record_batch_builder`.

This is only really an issue when buffer contents exceed max_batch_size, but I think I'd prefer a consistent overestimate (i.e. slightly less aggressive batching) over a consistent underestimate (i.e. data loss in the extreme case).

I've spent a good amount of time puzzling over this; maybe too much. I'm going to add another unit test for some of these assumptions, but I'm open to other suggestions as well.

Reply:
It's probably because most of the time the vint size from the `offset` is always zero when measuring this way, but as you accumulate a batch it's non-zero...

Agreed that over estimating is better. Maybe we just need a free function in `model` to do the calculation we want...

Reply:
>probably because...as you accumulate a batch it's non-zero

Ah, that sounds very plausible. Incidentally, I think that does mean that we'll reliably overestimate, and slightly better than the previous version at that. (prior estimate accounted for this by pessimizing `offset_delta` to vint::max_bytes).

>Maybe we just need a free function in model

That would be great, but I'm not sure what it would look like. We don't know how large the offset will grow until we're done, right?

Reply:
You could just use `model::offset::max()`

Honeslty here is my strawman...

```c++
// New constructor for record that computes the sizes
model::record tmp = model::record(
  key,
  value,
  /*headers=*/{},
  model::offset_delta::max(),
  /*ts_delta=*/std::numeric_limits<int32_t>::max());
size_t record_size = tmp.size_bytes();
roll_if_adding(record_size);
batch.add_raw_kv(record.release_key(), value.release_value());
```

Sorry for the back and forth. What do you think?

Reply:
I don't hate it. Keeping the `vint` calculations hidden feels like a significant improvement.

>Sorry for the back and forth

not at all. like I said, i've been waffling on this for a minute; better to externalize the discussion imo.